app/models/solidus_stripe/payment_method.rb

```diff
@@ -7,37 +7,143 @@ class PaymentMethod < ::Spree::PaymentMethod
+      def can_credit?(payment)
+        payment.completed? && payment.credit_allowed > 0
+      end
     end
 
-    def gateway_class
-      Gateway
+    concerning :Configuration do
+      def partial_name
+        "stripe"
+      end
+
+      alias cart_partial_name partial_name
+      alias product_page_partial_name partial_name
+      alias risky_partial_name partial_name
+
+      def source_required?
+        false
+      end
+
+      def payment_source_class
+        PaymentSource
+      end
+
+      def gateway_class
+        Gateway
+      end
     end
 
-    def create_profile(payment)
-      payment_intent = payment.source.payment_intent
+    concerning :Order do
+      def find_in_progress_payment_for(order)
+        payments = order.payments
+          .where.not(state: %w[completed invalid void]) # in_progress
+          .where(payment_method: self)
+          .order(:created_at)
+          .entries
+
+        *old_payments, payment = payments
+        old_payments.each(&:invalidate!)
+
+        if payment && payment.amount != order.total
+          payment.cancel!
+          payment = nil
+        end
+
+        payment
+      end
+
+      def create_in_progress_payment_for(order)
+        transaction do
+          intent = gateway.request do
+            Stripe::PaymentIntent.create({
+              amount: gateway.to_stripe_amount(
+                order.display_total.money.fractional,
+                order.currency,
+              ),
+              currency: order.currency,
 
-      if payment_intent && payment_intent.customer.blank?
-        payment.payment_method.gateway.request do
-          payment_intent.customer = Stripe::Customer.new email: payment.order.email
-          payment_intent.customer.save
-          payment_intent.save
+              # The capture method should stay manual in order to
+              # avoid capturing the money before the order is completed.
+              capture_method: 'manual',
+            })
+          end
+
+          order.payments
+            .create!(
+              payment_method: self,
+              response_code: intent.id,
+              amount: order.total,
+            ).tap(&:started_processing!)
         end
       end
 
-      self
+      def find_or_create_in_progress_payment_for(order)
+        payment = find_in_progress_payment_for(order)
+        intent = find_intent_for(payment) if payment
+
+        payment = nil if intent.nil?
+        payment = nil unless intent&.status == 'requires_payment_method'
+
+        payment ||= create_in_progress_payment_for(order)
+        payment
+      end
     end
 
-    def payment_profiles_supported?
-      true
+    concerning :Payment do
+      def create_profile(payment)
+        payment_intent = find_intent_for(payment)
+
+        if payment_intent && payment_intent.customer.blank?
+          payment.payment_method.gateway.request do
+            payment_intent.customer = Stripe::Customer.new email: payment.order.email
```

Comment:
From my tests, this does not work for several reasons:
- `Stripe::Customer.new` does not do anything.
- You cannot add a customer on an existing intent. You have to create the inetnt with the customer directly.
```bash
Stripe::InvalidRequestError: You cannot modify `customer` on a PaymentIntent once it already has been set. To fulfill a payment with a different Customer, cancel this PaymentIntent and create a new one.
from ~/.rvm/gems/ruby-2.7.7@epicery/gems/stripe-8.2.0/lib/stripe/stripe_client.rb:722:in `handle_error_response'
```
- You should not always create a new customer but find an existing customer if it exists. Something like:
```
            payment_intent.customer = find_or_create_customer(payment)
```
and then something like:
```ruby
      def find_or_create_customer(payment)
        res = payment.payment_method.gateway.request do
          Stripe::Customer.search({
            query: "email: '#{payment.order.email}'"
          }, stripe_version: '2020-08-27') # search is not supported for previous versions
        end
        if res[:data] && res[:data].first
          customer = res[:data].first
        else
          customer = payment.payment_method.gateway.request do
            Stripe::Customer.create({ email: payment.order.email })
          end
        end
        customer
      end
```

Reply:
also tracked in #172